### DISCUSSION TITLE
Dynamic binning of time-series data on zoom

### DISCUSSION THREAD

**Discussion by ibnesayeed** [vega/altair#2669] [Q&A]
Suppose, we have a time-series data of a web server access log. We want to render a bar plot of number of requests to the server per unit time. Suppose, we bind the interactivity of the chart so that it only responds on x-axis on zoom. Now, we would like for the bar plot to change the bin size (time unit) as we zoom in and out. Suppose, initially it was a daily count, on zooming in it should turn into an hourly bar plot (and per minute or per second on further zooming in), and weekly, monthly, and yearly on zooming out. The y-axis scale should change to reflect the changing heights of bars.

Is this something possible with Altair? If so, any examples or code samples for this behavior would be wonderful.

**Comment by joelostblom** [comment]
I am not aware of anyway to change the type of binning dynamically upon zooming. Vega-lite does some automatic formatting of the axes depending on the data in the chart, so you might be able to achieve something similar to what you want by using a `transform_filter` to narrow the data that is used for computing the axes labels.

**Comment by christian-johnson** [comment]
For anyone who stumbles across this discussion in the future, it turns out this is actually possible natively with altair, using `alt.selection_interval(bind="scales")` instead of `.interactive()` . Here's an example:

```
import altair as alt
from vega_datasets import data
import pandas as pd

source = data.flights_5k.url
df = pd.read_json(source)
df = df.groupby(pd.Grouper(key='date', freq='h')).size().reset_index(name='count')

base = alt.Chart(df).transform_calculate(
    time="datetime(year(datum.date), month(datum.date), date(datum.date), hours(datum.date))"
).mark_area(interpolate='step-after').encode(
    y='sum(count):Q',
).properties(
    width=600,
    height=200
)
zoom_selection = alt.selection_interval(bind="scales")

base.encode(
    alt.X('time:T',
        axis=alt.Axis(format="%b %d, %H:00"))
          .bin(maxbins=100, extent=zoom_selection)
      .scale(domain=zoom_selection)
  ).add_params(zoom_selection)
```


https://github.com/user-attachments/assets/290fc6ec-de25-4c13-964c-ee80d034ad15
